modules/apps/portal/portal-log4j-test/src/test/java/com/liferay/portal/log4j/Log4jTest.java

```diff
@@ -0,0 +1,464 @@
+import com.liferay.portal.log.Log4jLogFactoryImpl;
+import com.liferay.portal.util.PropsImpl;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.io.StringWriter;
+
+import java.net.URI;
+import java.net.URL;
+
+import java.text.SimpleDateFormat;
+
+import java.util.Date;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.log4j.SimpleLayout;
+import org.apache.log4j.WriterAppender;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * @author Hai Yu
+ */
+public class Log4jTest {
+
+	@BeforeClass
+	public static void setUpClass() throws Exception {
+		PropsUtil.setProps(new PropsImpl());
+
+		_originalOutputStream = System.out;
+
+		_printStream = new TeePrintStream(_baos, _originalOutputStream);
+
+		System.setOut(_printStream);
+
+		Thread currentThread = Thread.currentThread();
+
+		ClassLoader classLoader = currentThread.getContextClassLoader();
+
+		URL url = classLoader.getResource(
+			"com/liferay/portal/log4j/dependencies/log4j.xml");
+
+		String urlContent = "";
+
+		try (InputStream inputStream = url.openStream()) {
+			byte[] bytes = _getBytes(inputStream);
+
+			urlContent = new String(bytes, StringPool.UTF8);
+		}
+
+		String dir = StringUtil.replace(
+			System.getProperty("user.dir"), '\\', '/');
+
+		urlContent = StringUtil.replace(urlContent, "@user_dir@", dir);
+
+		_tempFile = new File(System.getProperty("user.dir"), "/log4j.xml");
+
+		try {
+			_tempFile.createNewFile();
+
+			DataOutputStream outstream = new DataOutputStream(
+				new FileOutputStream(_tempFile));
+
+			outstream.write(urlContent.getBytes());
+
+			outstream.flush();
+
+			outstream.close();
+
+			URI uri = _tempFile.toURI();
+
+			Log4JUtil.configureLog4J(uri.toURL());
+
+			LogFactoryUtil.setLogFactory(new Log4jLogFactoryImpl());
+		}
+		catch (IOException ioException) {
+			throw ioException;
+		}
+	}
+
+	@AfterClass
+	public static void tearDownClass() {
+		_printStream.flush();
+
+		_printStream.close();
+
+		System.setOut(_originalOutputStream);
+
+		LogManager.shutdown();
+
+		if (_tempFile != null) {
+			_tempFile.delete();
+		}
+
+		File logDir = new File(
+			StringUtil.replace(System.getProperty("user.dir"), '\\', '/'),
+			"logs");
+
+		for (File file : logDir.listFiles()) {
+			file.delete();
+		}
+
+		logDir.delete();
+	}
+
+	@Test
+	public void testConsoleAppender() {
+		Log log = LogFactoryUtil.getLog(Log4jTest.class);
+
+		log.info("Test Message");
+
+		String[] logMessages = StringUtil.splitLines(_baos.toString());
+
+		String expectedOutput = logMessages[logMessages.length - 1];
+
+		_assertLog(expectedOutput, 147);
```

Comment:
@yuhai there are 2 big issues here (also in you other output assertions).

1. You hardcoded the line number, so the test will fail everytime we change any code.
2. You use the output from the `_baos` as "expected", but it's actually what the logger prints, and what you need to check.

Please change your code to avoid asserting the log output with line numbers and exact output. For example,

1. Assert that the log output you get from the `_baos` meets the format you configured.
2. Assert correctness of the critical parts, like the logging level, the thread name, and the message.
3. Add more tests to test output on different levels, as well as different logging methods.

```
	public void info(Object msg);

	public void info(Object msg, Throwable t);

	public void info(Throwable t);
```

Reply:
The same applies to your `_assertXmlLog`.